File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/LogicalLogRecord.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using System;
    using System.Diagnostics.CodeAnalysis;
    using System.Fabric;
    using System.Fabric.Common;
    using System.IO;
    using System.Threading.Tasks;

    internal class LogicalLogRecord : LogRecord
    {
        internal const ulong InvalidLogicalRecordOffset = ulong.MaxValue;
        private const int SizeOnWire = sizeof(int) + sizeof(int) + sizeof(uint) + sizeof(long);

        internal static readonly LogicalLogRecord InvalidLogicalLogRecord = new LogicalLogRecord();
        private static readonly uint ApproximateDiskSpaceUsed = sizeof(int);

        private Task<long> replicatorTask;
        private ArraySegment<byte> replicatedData;
        private BinaryWriter replicatedDataWriter;

        internal LogicalLogRecord(LogRecordType recordType, ulong recordPosition, long lsn)
            : base(recordType, recordPosition, lsn)
        {
            this.replicatorTask = null;
        }

        internal LogicalLogRecord(LogRecordType recordType)
            : base(recordType)
        {
            this.replicatorTask = null;
        }

        protected LogicalLogRecord()
            : base()
        {
            this.replicatorTask = null;
        }

        internal virtual bool IsLatencySensitiveRecord
        {
            get { return false; }
        }

        internal bool IsReplicated
        {
            get
            {
                if (this.replicatorTask != null)
                {
                    return this.replicatorTask.IsCompleted;
                }

                return true;
            }
        }

        internal Task<long> ReplicatorTask
        {
            get { return this.replicatorTask; }

            set { this.replicatorTask = value; }
        }

        internal async Task AwaitReplication()
        {
            if (this.replicatorTask != null)
            {
                await this.replicatorTask.ConfigureAwait(false);
            }

            return;
        }

        internal virtual void InvalidateReplicatedDataBuffers(SynchronizedBufferPool<BinaryWriter> pool)
        {
            if (pool != null && this.replicatedDataWriter != null)
            {
                pool.Return(this.replicatedDataWriter);
            }

            this.replicatedDataWriter = null;
            this.replicatedData = new ArraySegment<byte>();
        }

        [SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1126:PrefixCallsCorrectly", Justification = "Reviewed.")]
        internal OperationData ToOperationData(BinaryWriter bw)
        {
            this.replicatedDataWriter = bw;
            this.replicatedDataWriter.BaseStream.Position = 0;
            var operationData = new OperationData();
            this.Write(this.replicatedDataWriter, operationData, false, false);
            return operationData;
        }

        protected virtual int GetSizeOnWire()
        {
            return SizeOnWire;
        }

        protected override void Read(BinaryReader br, bool isPhysicalRead)
        {
            base.Read(br, isPhysicalRead);
            this.ReadPrivate(br, isPhysicalRead);

            return;
        }

        protected override void ReadLogical(OperationData operationData, ref int index)
        {
            base.ReadLogical(operationData, ref index);

            using (var br = new BinaryReader(IncrementIndexAndGetMemoryStreamAt(operationData, ref index)))
            {
                this.ReadPrivate(br, false);
            }

            return;
        }

        protected override void Write(BinaryWriter bw, OperationData operationData, bool isPhysicalWrite, bool forceRecomputeOffsets)
        {
            base.Write(bw, operationData, isPhysicalWrite, forceRecomputeOffsets);
            var startingPosition = bw.BaseStream.Position;

            if (this.replicatedData.Array == null)
            {
                // Metadata Size
                bw.BaseStream.Position += sizeof(int);

                // Future fields

                // End of metadata.
                var endPosition = bw.BaseStream.Position;
                var sizeOfSection = checked((int) (endPosition - startingPosition));
                bw.BaseStream.Position = startingPosition;
                bw.Write(sizeOfSection);
                bw.BaseStream.Position = endPosition;

                this.replicatedData = CreateArraySegment(startingPosition, bw);
            }

            operationData.Add(this.replicatedData);
        }

        private void ReadPrivate(BinaryReader br, bool isPhysicalRead)
        {
            // Read Metadata size.
            var logicalStartingPosition = br.BaseStream.Position;
            var logicalSizeOfSection = br.ReadInt32();
            var logicalEndPosition = logicalStartingPosition + logicalSizeOfSection;

            // Read Metadata fields (future)

            // Jump to the end of the section ignoring fields that are not understood.
            Utility.Assert(logicalEndPosition >= br.BaseStream.Position, "Could not have read more than section size.");
            br.BaseStream.Position = logicalEndPosition;

            this.UpdateApproximateDiskSize();
        }

        private void UpdateApproximateDiskSize()
        {
            this.ApproximateSizeOnDisk += ApproximateDiskSpaceUsed;
        }
    }
}

File: src/prod/src/data/txnreplicator/logrecordlib/LogicalLogRecord.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace ktl;
using namespace Data::LogRecordLib;
using namespace TxnReplicator;
using namespace Data::Utilities;

const ULONG LogicalLogRecord::DiskSpaceUsed = sizeof(LONG32);
const ULONG LogicalLogRecord::SizeOnWire = sizeof(LONG32) + sizeof(LONG32) + sizeof(ULONG32) + sizeof(LONG64);

LogicalLogRecord::LogicalLogRecord()
    : LogRecord()
    , replicationTask_()
    , replicatedData_()
    , replicatedDataBinaryWriter_(GetThisAllocator())
{
}

LogicalLogRecord::LogicalLogRecord(
    __in LogRecordType::Enum recordType,
    __in PhysicalLogRecord & invalidPhysicalLogRecord)
    : LogRecord(recordType, invalidPhysicalLogRecord)
    , replicationTask_()
    , replicatedData_(nullptr)
    , replicatedDataBinaryWriter_(GetThisAllocator())
{
    UpdateApproximateDiskSize(); 
}

LogicalLogRecord::LogicalLogRecord(
    __in LogRecordType::Enum recordType,
    __in ULONG64 recordPosition,
    __in LONG64 lsn,
    __in PhysicalLogRecord & invalidPhysicalLogRecord)
    : LogRecord(recordType, recordPosition, lsn, invalidPhysicalLogRecord)
    , replicationTask_()
    , replicatedData_(nullptr)
    , replicatedDataBinaryWriter_(GetThisAllocator())
{
}

LogicalLogRecord::~LogicalLogRecord()
{
    if (replicationTask_)
    {
        ASSERT_IFNOT(
            replicationTask_->IsCompleted,
            "Replication task for LSN {0} is not complete during destruction",
            Lsn);
    }
}

bool LogicalLogRecord::IsAllowedToThrottle(__in LogicalLogRecord const & logicalRecord)
{
    LogRecordType::Enum recordType = logicalRecord.RecordType;

    return
        recordType == LogRecordType::Enum::BeginTransaction ||
        recordType == LogRecordType::Enum::Operation;
}

void LogicalLogRecord::UpdateApproximateDiskSize()
{
    ApproximateSizeOnDisk = ApproximateSizeOnDisk + DiskSpaceUsed;
}

std::wstring LogicalLogRecord::ToString() const
{
    return __super::ToString();
}

OperationData::CSPtr LogicalLogRecord::SerializeLogicalData()
{
    OperationData::SPtr operationData = OperationData::Create(GetThisAllocator());
    Write(replicatedDataBinaryWriter_, *operationData, false, false);

    OperationData::CSPtr returnData(operationData.RawPtr());

    return returnData;
}

void LogicalLogRecord::ReleaseSerializedLogicalDataBuffers()
{
    replicatedDataBinaryWriter_.Reset();
}

Awaitable<NTSTATUS> LogicalLogRecord::AwaitReplication()
{
    ASSERT_IF(Lsn <= 0, "Invalid Replication Lsn in AwaitReplication")
    NTSTATUS status = co_await replicationTask_->AwaitCompletion();

    co_return status;
}

void LogicalLogRecord::Read(
    __in BinaryReader & binaryReader,
    __in bool isPhysicalRead)
{
    __super::Read(binaryReader, isPhysicalRead);
    ReadPrivate(binaryReader, isPhysicalRead);
}

void LogicalLogRecord::ReadLogical(
    __in OperationData const & operationData,
    __inout INT & index)
{
    __super::ReadLogical(operationData, index);
    
    KBuffer::CSPtr buffer = OperationData::IncrementIndexAndGetBuffer(operationData, index);
    BinaryReader reader(*buffer, GetThisAllocator());
    
    ReadPrivate(reader, false);
}

void LogicalLogRecord::Write(
    __in BinaryWriter & binaryWriter,
    __inout OperationData & operationData,
    __in bool isPhysicalWrite,
    __in bool forceRecomputeOffsets)
{
    __super::Write(binaryWriter, operationData, isPhysicalWrite, forceRecomputeOffsets);

    if (replicatedData_ == nullptr)
    {
        ULONG32 startingPosition = binaryWriter.Position;

        // Metadata Size
        binaryWriter.Position += sizeof(ULONG32);

        // Future fields

        // End of metadata.
        ULONG32 endPosition = binaryWriter.Position;
        ULONG32 sizeOfSection = endPosition - startingPosition;

        binaryWriter.Position = startingPosition;
        binaryWriter.Write(sizeOfSection);
        binaryWriter.Position = endPosition;

        replicatedData_ = binaryWriter.GetBuffer(startingPosition);
    }

    operationData.Append(*replicatedData_);
}

ULONG LogicalLogRecord::GetSizeOnWire() const
{
    return SizeOnWire;
}

void LogicalLogRecord::ReadPrivate(
    __in BinaryReader & binaryReader,
    __in bool isPhysicalRead)
{
    UNREFERENCED_PARAMETER(isPhysicalRead);

    // Read Metadata size
    ULONG32 logicalStartingPosition = binaryReader.Position;
    ULONG32 logicalSizeOfSection = 0;
    binaryReader.Read(logicalSizeOfSection);
    ULONG32 logicalEndPosition = logicalStartingPosition + logicalSizeOfSection;

    // Read Metadata fields (future)

    // Jump to the end of the section ignoring fields that are not understood.
    ASSERT_IFNOT(
        logicalEndPosition >= binaryReader.Position,
        "Invalid logical log record, end position: {0} reader position:{1}",
        logicalEndPosition, binaryReader.Position);

    binaryReader.Position = logicalEndPosition;

    UpdateApproximateDiskSize();
}

bool LogicalLogRecord::Test_Equals(__in LogRecord const & other) const
{
    return __super::Test_Equals(other);
}
